api/v2api/full.go

```diff
@@ -0,0 +1,142 @@
+package v2api
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-jsonrpc"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -destination=v2mocks/mock_full.go -package=v2mocks . FullNode
+
+type FullNode interface {
+	// ChainNotify returns channel with chain head updates. Takes an optional
+	// EpochSelector argument to specify the point in the chain to follow, which
+	// can be either "latest" or "finalized". This defaults to "latest" if not
+	// specified.
+	//
+	// The first message is guaranteed to be of length == 1, and type ==
+	// "current".
+	//
+	// The optional argument is decoded as a types.OptionalEpochSelectorArg.
+	ChainNotify(ctx context.Context, p jsonrpc.RawParams) (<-chan []*api.HeadChange, error) //perm:read
```

Comment:
Just so I learn: how popular/critical is it to have a "push" notification API? e.g. what would immediately break if this is converted to a pull model?

Reply:
Anything that returns a channel is usable via WebSockets so push data to you, otherwise you're polling, probably with `ChainHead`, but the nice thing about `ChainNotify` is that you get a constant stream of changes so you have a consistent chain including reverts and applies as it reorgs so you just process them as they come rather than trying to figure it out.

The alternative of polling `ChainHead` you'd get different tipsets when you poll and have to react to those changes in whatever way your local state demands. If you want to replicate the full `ChainNotify` behaviour you might need to follow up with `ChainGetPath` to get the changes between what you currently know in your local state as the chain head and what `ChainHead` is telling you what it is now, that way you could rebuild a `ChainNotify` but it'd be gnarly.

`ChainNotify` is a pretty critical API internally for Lotus, and there's at least one place that lotus-miner uses it (and therefore likely Curio too) and I wouldn't be surprised to see other external applications we're not aware of.

Reply:
Thanks, Rod.

Is there anything that inherently depends on the push model though? For example, is there something critical about the end-to-end latency that `ChainNotify` users care about/cannot be provided in a "pull" model?

I am specifically interested in its criticality for full node API users (i.e. excluding internal use by lotus node itself).

Reply:
I'm not sure. It's certainly a handy API. You can spin it up on a WebSocket connection and react to the changes as they come in.

I'm just not sure how many users are reliant on changes right at the head, I try and avoid it where possible because it's chaos! Maybe exchanges care? I don't really understand the certainty model that exchanges operate under.

You can even use `ChainNotify` just for a singular "tell me when the next epoch has arived" so I can get on with something, that's a pretty typical use internally and I can imagine that being useful externally too. I just can't point to an external use-case for it.

https://github.com/filecoin-project/lotus/blob/1b276e402caeb0f139bd881d76cdcdd206221f55/itests/fevm_test.go#L1409-L1413

Reply:
The context for my silly questions: It's my sense that such push APIs are 1) hard to get correct for external use, and 2) hard to scale (e.g. in the usecase of exchanges, archival nodes etc.).

Hence the conversation on whether we can avoid this API entirely in v2, in favour of a "Smart Client" implementation that affectively offers the same functionality but with a far smaller maintenance/correctness burden for Lotus Node implementation.